tutorials/docs-17-implementing-samplers/index.qmd

```diff
@@ -0,0 +1,479 @@
+
+```{julia}
+# Perform 1000 iterations with our `MALA` sampler.
+samples = sample(model_with_grad, sampler, 10_000; initial_state=state, progress=false)
+# Concatenate into a matrix.
+samples_matrix = stack(samples)
+```
+
+```{julia}
+# Compute the marginal means and standard deviations.
+hcat(mean(samples_matrix; dims=2), std(samples_matrix; dims=2))
+```
+
+Let's visualize the samples
+
+```{julia}
+using StatsPlots
+plot(transpose(samples_matrix[:, 1:10:end]), alpha=0.5, legend=false)
+```
+
+Look at that! Things are working; amazin'.
+
+We can also exploit [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl)'s parallel sampling capabilities:
+
+```{julia}
+# Run separate 4 chains for 10 000 iterations using threads to parallelize.
+num_chains = 4
+samples = sample(
+    model_with_grad,
+    sampler,
+    MCMCThreads(),
+    10_000,
+    num_chains;
+    # Note we need to provide an initial state for every chain.
+    initial_state=fill(state, num_chains),
+    progress=false
+)
+samples_array = stack(map(stack, samples))
+```
+
+But the fact that we have to provide the `AbstractMCMC.sample` call, etc. with an `initial_state` to get started is a bit annoying. We can avoid this by also defining a `AbstractMCMC.step` *without* the `state` argument:
+
+```{julia}
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model_wrapper::AbstractMCMC.LogDensityModel,
+    sampler::MALA;
+    # NOTE: No state provided!
+    kwargs...
+)
+    model = model_wrapper.logdensity
+    # Let's just create the initial state by sampling using  a Gaussian.
+    x = randn(rng, LogDensityProblems.dimension(model))
+
+    return x, MALAState(x)
+end
+```
+
+Equipped with this, we no longer need to provide the `initial_state` everywhere:
+
+```{julia}
+samples = sample(model_with_grad, sampler, 10_000; progress=false)
+samples_matrix = stack(samples)
+hcat(mean(samples_matrix; dims=2), std(samples_matrix; dims=2))
+```
+
+## Using our sampler with Turing.jl
+
+As we promised, all of this hassle of implementing our `MALA` sampler in a way that uses [LogDensityProblems.jl](https://github.com/tpapp/LogDensityProblems.jl) and [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl) gets us something more than *just* an "automatic" implementation of `AbstractMCMC.sample`.
+
+It also enables use with Turing.jl through the `externalsampler`, but we need to do one final thing first: we need to tell Turing.jl how to extract a vector of parameters from the "sample" returned in our implementation of `AbstractMCMC.step`. In our case, the "sample" is just a `Vector{<:Real}`, so we just need the following line:
+
+```{julia}
+# Load Turing.jl.
+using Turing
+
+# Overload the `getparams` method for our "sample" type, which is just a vector.
+Turing.Inference.getparams(::Turing.Model, x::AbstractVector{<:Real}) = x
```

Comment:
I'm a bit confused by this, which looks like type piracy to me. Should this rather be defined for `MALAState`?

EDIT: Also just noticed that the docstring of `getparams` says "Return a named tuple of parameters", which adds to my confusion.